File: benchmarks/cross/cases/data/byte_scan.py
import sys


SEGMENT_SIZE = 32_768
SEGMENT_COUNT = 256
CHECKSUM_MULTIPLIER = 1_000_003
CHECKSUM_MODULUS = 1_000_000_007
DELIMITER = b"|"


def requested_work() -> int:
    if len(sys.argv) < 2:
        return 1
    try:
        return max(0, int(sys.argv[1]))
    except ValueError:
        return 1


def make_input() -> bytes:
    segment = b"a" * (SEGMENT_SIZE - 1) + DELIMITER
    return segment * SEGMENT_COUNT


def scan(input_bytes: bytes) -> int:
    start = 0
    total = 0
    while start < len(input_bytes):
        index = input_bytes.find(DELIMITER, start)
        if index < 0:
            break
        total += index + 1
        start = index + 1
    return total


def main() -> None:
    input_bytes = make_input()
    checksum = 0
    for _ in range(requested_work()):
        checksum = (
            checksum * CHECKSUM_MULTIPLIER + scan(input_bytes)
        ) % CHECKSUM_MODULUS
    print(checksum)


if __name__ == "__main__":
    main()

File: benchmarks/cross/cases/data/byte_scan.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"strconv"
)

const (
	segmentSize        = 32768
	segmentCount       = 256
	inputSize          = segmentSize * segmentCount
	delimiter          = byte('|')
	checksumMultiplier = uint64(1000003)
	checksumModulus    = uint64(1000000007)
)

func requestedWork() uint64 {
	if len(os.Args) < 2 {
		return 1
	}
	work, err := strconv.ParseInt(os.Args[1], 10, 64)
	if err != nil {
		return 1
	}
	if work < 0 {
		return 0
	}
	return uint64(work)
}

func makeInput() []byte {
	segment := bytes.Repeat([]byte{'a'}, segmentSize)
	segment[len(segment)-1] = delimiter
	input := make([]byte, 0, inputSize)
	for i := 0; i < segmentCount; i++ {
		input = append(input, segment...)
	}
	return input
}

func scan(input []byte) uint64 {
	from := 0
	var total uint64
	for from < len(input) {
		offset := bytes.IndexByte(input[from:], delimiter)
		if offset < 0 {
			break
		}
		index := from + offset
		total += uint64(index + 1)
		from = index + 1
	}
	return total
}

func main() {
	work := requestedWork()
	input := makeInput()
	var checksum uint64
	for rep := uint64(0); rep < work; rep++ {
		checksum = (checksum*checksumMultiplier + scan(input)) % checksumModulus
	}
	fmt.Println(checksum)
}
